common/archiver/s3store/visibilityArchiver.go

```diff
@@ -0,0 +1,252 @@
+// Copyright (c) 2019 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package s3store
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3iface"
+
+	"github.com/uber/cadence/.gen/go/shared"
+	"github.com/uber/cadence/common/archiver"
+	"github.com/uber/cadence/common/log/tag"
+	"github.com/uber/cadence/common/service/config"
+)
+
+type (
+	visibilityArchiver struct {
+		container   *archiver.VisibilityBootstrapContainer
+		s3cli       s3iface.S3API
+		queryParser QueryParser
+	}
+
+	visibilityRecord archiver.ArchiveVisibilityRequest
+
+	queryVisibilityRequest struct {
+		domainID      string
+		pageSize      int
+		nextPageToken []byte
+		parsedQuery   *parsedQuery
+	}
+)
+
+const (
+	errEncodeVisibilityRecord = "failed to encode visibility record"
+	indexKeyStartTimeout      = "startTimeout"
+	indexKeyCloseTimeout      = "closeTimeout"
+)
+
+// NewVisibilityArchiver creates a new archiver.VisibilityArchiver based on s3
+func NewVisibilityArchiver(
+	container *archiver.VisibilityBootstrapContainer,
+	config *config.S3Archiver,
+) (archiver.VisibilityArchiver, error) {
+	return newVisibilityArchiver(container, config)
+}
+
+func newVisibilityArchiver(
+	container *archiver.VisibilityBootstrapContainer,
+	config *config.S3Archiver) (*visibilityArchiver, error) {
+	s3Config := &aws.Config{
+		Endpoint:         config.Endpoint,
+		Region:           aws.String(config.Region),
+		S3ForcePathStyle: aws.Bool(config.S3ForcePathStyle),
+		MaxRetries:       aws.Int(0),
+	}
+	sess, err := session.NewSession(s3Config)
+	if err != nil {
+		return nil, err
+	}
+	return &visibilityArchiver{
+		container:   container,
+		s3cli:       s3.New(sess),
+		queryParser: NewQueryParser(),
+	}, nil
+}
+
+func (v *visibilityArchiver) Archive(
```

Comment:
Please emit metrics inside this method.

MetricScope: `VisibilityArchiverScope`
MetricsNames:
```
        CadenceLatency
	VisibilityArchiverArchiveNonRetryableErrorCount
	VisibilityArchiverArchiveTransientErrorCount
	VisibilityArchiveSuccessCount
```

Reply:
Done!